src/v/cluster/distributed_kv_stm.h

```diff
@@ -268,7 +269,33 @@ class distributed_kv_stm final : public raft::persisted_stm<> {
             co_return errc::success;
         }
         co_return co_await replicate_and_wait(
-          make_kv_data_batch_remove_key<Key, Value>(key));
+          make_kv_data_batch_remove_all<Key, Value>({key}));
+    }
+
+    ss::future<errc> remove_all(ss::noncopyable_function<bool(Key)> pred) {
+        auto holder = _gate.hold();
+        auto units = co_await _snapshot_lock.hold_read_lock();
```

Comment:
I feel like this could hold a write lock since remove_all seems like a rare operation and the side effects of it are easier to explain if the entire operation sees a consistent snapshot of the state.

Reply:
Yeah it is a rare case, so probably cleaner to make it see consistent state. I've pushed an update.